x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/QueryList.java

```diff
@@ -50,55 +62,163 @@ int getPositionCount() {
+            this.searchExecutionContext = searchExecutionContext;
+            this.inputDataType = inputDataType;
+            this.blockValueReader = blockToJavaObject();
+        }
+
+        @Override
+        Query getQuery(int position) {
+            final int first = block.getFirstValueIndex(position);
+            final int count = block.getValueCount(position);
+            return switch (count) {
+                case 0 -> null;
+                case 1 -> field.termQuery(blockValueReader.apply(first), searchExecutionContext);
+                default -> {
+                    final List<Object> terms = new ArrayList<>(count);
+                    for (int i = 0; i < count; i++) {
+                        final Object value = blockValueReader.apply(first + i);
+                        terms.add(value);
                     }
-                };
-            }
-        };
+                    yield field.termsQuery(terms, searchExecutionContext);
+                }
+            };
+        }
+
+        private IntFunction<Object> blockToJavaObject() {
+            return switch (block.elementType()) {
+                case BOOLEAN -> {
+                    BooleanBlock booleanBlock = (BooleanBlock) block;
+                    yield booleanBlock::getBoolean;
+                }
+                case BYTES_REF -> {
+                    BytesRefBlock bytesRefBlock = (BytesRefBlock) block;
+                    if (inputDataType == IP) {
+                        yield offset -> {
+                            bytesRefBlock.getBytesRef(offset, scratch);
+                            if (ipBytes.length != scratch.length) {
+                                // Lucene only support 16-byte IP addresses, even IPv4 is encoded in 16 bytes
+                                throw new IllegalStateException("Cannot decode IP field from bytes of length " + scratch.length);
+                            }
+                            System.arraycopy(scratch.bytes, scratch.offset, ipBytes, 0, scratch.length);
+                            InetAddress ip = InetAddressPoint.decode(ipBytes);
+                            return ip;
+                        };
+                    }
+                    yield offset -> bytesRefBlock.getBytesRef(offset, new BytesRef());
+                }
+                case DOUBLE -> {
+                    DoubleBlock doubleBlock = ((DoubleBlock) block);
+                    yield doubleBlock::getDouble;
+                }
+                case INT -> {
+                    IntBlock intBlock = (IntBlock) block;
+                    yield intBlock::getInt;
+                }
+                case LONG -> {
+                    LongBlock longBlock = (LongBlock) block;
+                    if (field instanceof RangeFieldMapper.RangeFieldType rangeFieldType && rangeFieldType.rangeType() == RangeType.DATE) {
+                        yield offset -> rangeFieldType.dateTimeFormatter().formatMillis(longBlock.getLong(offset));
```

Comment:
No, the range query does something weird with the millis, setting `max=min-1` when passed a single value, which then fails the validation. Perhaps if I explicitly pass in a `max=min+1` it will bypass this?